src/test/ui/rfc-2008-non-exhaustive/uninhabited.rs

```diff
@@ -0,0 +1,59 @@
+// aux-build:uninhabited.rs
+// compile-pass
+#![deny(unreachable_patterns)]
+#![feature(exhaustive_patterns)]
+
+extern crate uninhabited;
+
+use uninhabited::{
+    PartiallyInhabitedVariants,
+    UninhabitedEnum,
+    UninhabitedStruct,
+    UninhabitedTupleStruct,
+    UninhabitedVariants,
+};
+
+fn uninhabited_enum() -> Option<UninhabitedEnum> {
+    None
+}
+
+fn uninhabited_variant() -> Option<UninhabitedVariants> {
+    None
+}
+
+fn partially_inhabited_variant() -> PartiallyInhabitedVariants {
+    PartiallyInhabitedVariants::Tuple(3)
+}
+
+fn uninhabited_struct() -> Option<UninhabitedStruct> {
+    None
+}
+
+fn uninhabited_tuple_struct() -> Option<UninhabitedTupleStruct> {
+    None
+}
+
+// This test checks that non-exhaustive types that would normally be considered uninhabited within
+// the defining crate are not considered uninhabited from extern crates.
```

Comment:
This relies on the assumption that the same logic for unreachable types is used for `unreachable_patterns` (which is a lint and not part of the type system).

To check the behavior fully, extend the tests with a `compile-fail` file:
```rust
extern crate uninhabited;

use uninhabited::{
    PartiallyInhabitedVariants,
    UninhabitedEnum,
    UninhabitedStruct,
    UninhabitedTupleStruct,
    UninhabitedVariants,
};

struct A;

// This is not allowed today but non-! but we want to test it for redundancy.
fn cannot_coerce_empty_enum_to_anything(x: UninhabitedEnum) -> A {
    x // ERROR
}

// This is caught by the normal need for `_` for `#[non_exhaustive]`
// but we also test it here for redundancy.
fn cannot_empty_match_empty_enum_to_anything(x: UninhabitedEnum) -> A {
    match x {} // ERROR
}

// rinse & repeat for the other types.
```

Reply:
I've added a few tests like these - I couldn't come up with too many more than you listed that I thought made sense (ie. had a different behaviour w/out non-exhaustive or when in the same crate vs extern crate). I did resolve a issue or two in adding those that you listed though.

Reply:
Added a note below :)